meter.go

```diff
@@ -0,0 +1,128 @@
+package metrics
+
+import (
+	"net/http"
+	"sync"
+	"time"
+
+	"gopkg.in/vinxi/layer.v0"
+)
+
+// PublishInterval defines the amount of time to wait between metrics publish cycles.
+// Defaults to 15 seconds.
+var PublishInterval = 15 * time.Second
+
+// MeterFunc represents the function interface to be implemented by metrics meter functions.
+type MeterFunc func(*Info, *Metrics)
+
+// Reporter represents the function interface to be implemented by metrics reporters.
+// Metric reporters are responsable of reading, filtering and adapting metrics data.
+// Also, reporters tipically sends the metrics to an external service.
+type Reporter interface {
+	Report(Report)
+}
+
+// Meter provides a metrics instrumentation for vinxi
+// Supports configurable metrics reporters and meters.
+type Meter struct {
+	sync.Mutex
+	quit      chan bool
+	meters    []MeterFunc
+	reporters []Reporter
+	metrics   *Metrics
+	runtime   *RuntimeCollector
+}
+
+// New creates a new metrics meter middleware.
+func New(l ...Reporter) *Meter {
+	m := &Meter{
+		reporters: l,
+		meters:    Meters,
+		metrics:   NewMetrics(),
+		quit:      make(chan bool),
+	}
+
+	// Bind the Go runtime stats collector
+	m.runtime = NewRuntimeCollector(m.gaugeRuntime)
+
+	// Start collector goroutines
+	go m.runtime.Start()
+	go m.Start()
+
+	return m
+}
+
+// AddMeter adds one or multiple meter functions.
+func (m *Meter) AddMeter(meters ...MeterFunc) {
+	m.Lock()
+	m.meters = append(m.meters, meters...)
+	m.Unlock()
+}
+
+// SetMeters sets a new set of meter functions, replacing the existent ones.
+func (m *Meter) SetMeters(meters []MeterFunc) {
+	m.Lock()
+	m.meters = meters
+	m.Unlock()
+}
+
+// AddReporter adds one or multiple metrics reporters.
+func (m *Meter) AddReporter(reporters ...Reporter) {
+	m.Lock()
+	m.reporters = append(m.reporters, reporters...)
+	m.Unlock()
+}
+
+// Register registers the metrics middleware function.
+func (m *Meter) Register(mw layer.Middleware) {
+	mw.UsePriority("request", layer.TopHead, m.measureHTTP)
+}
+
+// Publish release store.
+func (m *Meter) Publish() {
+	for _, reporter := range m.reporters {
+		reporter.Report(m.metrics.Snapshot())
```

Comment:
The same I described before. Initially I thought would be better to provide a side effects free `map` to the reporters, but this could be reverted. I'm not sure how to design to avoid nested issues in third-party developers.